main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -33,303 +33,406 @@ object Safety {
     //
     // Collect all errors.
     //
-    val defErrors = root.defs.flatMap {
+    val defsVal = traverse(root.defs) {
       case (_, defn) => visitDef(defn, root)
     }
 
-    val instanceErrors = visitSendable(root)
-
-    val errors = defErrors ++ instanceErrors
-
     //
     // Check if any errors were detected.
     //
-    if (errors.isEmpty)
-      root.toSuccess
-    else
-      Validation.Failure(errors.to(LazyList))
+    val sendableVal = visitSendable(root)
+
+    mapN(defsVal, sendableVal)((_, _) => root)
   }
 
   /**
     * Checks that no type parameters for types that implement `Sendable` of kind `Region`
     */
-  private def visitSendable(root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitSendable(root: Root)(implicit flix: Flix): Validation[List[Instance], CompilationMessage] = {
 
     val sendableClass = new Symbol.ClassSym(Nil, "Sendable", SourceLocation.Unknown)
-
-    root.instances.getOrElse(sendableClass, Nil) flatMap {
-      case Instance(_, _, _, _, tpe, _, _, _, loc) =>
-        if (tpe.typeArguments.exists(_.kind == Kind.Bool))
-          List(SafetyError.SendableError(tpe, loc))
-        else
-          Nil
+    traverse(root.instances.getOrElse(sendableClass, Nil)) {
+      case Instance(doc, ann, mod, clazz, tpe, tconstrs, defs, ns, loc) =>
+        if (tpe.typeArguments.exists(_.kind == Kind.Bool)) {
+          SafetyError.SendableError(tpe, loc).toFailure
+        } else {
+          Instance(doc, ann, mod, clazz, tpe, tconstrs, defs, ns, loc).toSuccess
+        }
     }
   }
 
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = {
```

Comment:
This should be returning a `Def`, not a `Root`

Reply:
Done